shopinvader_api_address/routers/address_service.py

```diff
@@ -120,49 +126,122 @@ def get_delivery_address(
-    # function of res.partner to do some checks before writing.
-    # These checks need more rights than what we are giving to
-    # the enspoint's user
-    # (e.g. snailmail/models/res_partner.py)
-    partner.sudo()._delete_shopinvader_delivery_address(address_id)
+    helper = env["shopinvader_api_address.address_router.helper"].new(
+        {"partner": partner}
+    )
+    helper._delete_address(address_id, "delivery")
+
+
+AddressType = Literal["invoicing", "delivery"]
+
+
+class ShopinvaderApiAddressRouterHelper(models.AbstractModel):
+    _name = "shopinvader_api_address.address_router.helper"
+    _description = "ShopInvader API Address Router Helper"
+
+    partner = fields.Many2one(
+        comodel_name="res.partner",
+    )
+
+    def _get_address(self, address_id: int, address_type: AddressType):
+        self.ensure_one()
+        return getattr(self.partner, f"_get_shopinvader_{address_type}_address")(
+            address_id
+        )
+
+    def _prepare_create_address_vals(
+        self,
+        data: InvoicingAddressCreate | DeliveryAddressCreate,
+        address_type: AddressType,
+    ):
+        return data.to_res_partner_vals()
+
+    def _prepare_update_address_vals(
+        self,
+        data: InvoicingAddressUpdate | DeliveryAddressUpdate,
+        address_type: AddressType,
+        address: ResPartner,
+    ):
+        return data.to_res_partner_vals()
+
+    def _create_address(
+        self,
+        data: InvoicingAddressCreate | DeliveryAddressCreate,
+        address_type: AddressType,
+    ):
+        vals = self._prepare_create_address_vals(data, address_type)
+        return getattr(self.partner, f"_create_shopinvader_{address_type}_address")(
+            vals
+        )
+
+    def _update_address(
+        self,
+        data: InvoicingAddressUpdate | DeliveryAddressUpdate,
+        address_type: AddressType,
+        address_id: int,
+    ):
+        address = self._get_address(address_id, address_type)
+        vals = self._prepare_update_address_vals(data, address_type, address)
+        # sudo() is needed because some addons override the write
+        # function of res.partner to do some checks before writing.
+        # These checks need more rights than what we are giving to
+        # the enspoint's user
+        # (e.g. snailmail/models/res_partner.py)
+        return getattr(
+            self.partner.sudo(), f"_update_shopinvader_{address_type}_address"
+        )(vals, address)
+
+    def _delete_address(
+        self,
+        address_id: int,
+        address_type: AddressType,
+    ):
+        address = self._get_address(address_id, address_type)
+        # sudo() is needed because some addons override the write
+        # function of res.partner to do some checks before writing.
+        # These checks need more rights than what we are giving to
+        # the enspoint's user
+        # (e.g. snailmail/models/res_partner.py)
+        return getattr(
```

Comment:
We should never return something from a method called with `sudo` it's a potential security breach